lib/Common/Memory/PageAllocator.cpp

```diff
@@ -548,36 +557,57 @@ PageSegmentBase<T>::DecommitPages(__in void * address, uint pageCount)
     if (!onlyUpdateState)
     {
 #pragma warning(suppress: 6250)
-        this->GetAllocator()->GetVirtualAllocator()->Free(address, pageCount * AutoSystemInfo::PageSize, MEM_DECOMMIT);
+        this->GetAllocator()->GetVirtualAllocator()->Free(address,
+          pageCount * AutoSystemInfo::PageSize, MEM_DECOMMIT);
     }
 
     Assert(decommitPageCount == (uint)this->GetCountOfDecommitPages());
 }
 
 template<typename T>
+void
+PageSegmentBase<T>::DecommitFreePagesInternal(uint index, uint pageCount)
+{
+    Assert(pageCount > 0 && (index + pageCount) <= this->GetAvailablePageCount());
+    char * currentAddress = this->address + (index * AutoSystemInfo::PageSize);
+    this->ClearRangeInFreePagesBitVector(index, pageCount);
+    this->SetRangeInDecommitPagesBitVector(index, pageCount);
+#pragma warning(suppress: 6250)
+    this->GetAllocator()->GetVirtualAllocator()->Free(currentAddress,
+      pageCount * AutoSystemInfo::PageSize, MEM_DECOMMIT);
+}
+
+template<typename T>
 size_t
 PageSegmentBase<T>::DecommitFreePages(size_t pageToDecommit)
 {
-    Assert(pageToDecommit != 0);
-    char * currentAddress = this->address;
+    Assert(pageToDecommit != 0 && this->GetAvailablePageCount() > 0);
 
-    uint decommitCount = 0;
-    for (uint i = 0; i < this->GetAvailablePageCount(); i++)
+    uint startIndex = 0, index = 0, decommitCount = 0;
+    do
     {
-        if (this->TestInFreePagesBitVector(i))
+        if (!this->TestInFreePagesBitVector(index))
         {
-            this->ClearBitInFreePagesBitVector(i);
-            this->SetBitInDecommitPagesBitVector(i);
-#pragma warning(suppress: 6250)
-            this->GetAllocator()->GetVirtualAllocator()->Free(currentAddress, AutoSystemInfo::PageSize, MEM_DECOMMIT);
-            decommitCount++;
+            if (startIndex < index)
+            {
+                uint pageCount = index - startIndex;
+                this->DecommitFreePagesInternal(startIndex, pageCount);
+            }
+            startIndex = index + 1;
         }
-        currentAddress += AutoSystemInfo::PageSize;
-        if (decommitCount == pageToDecommit)
+        else
         {
-            break;
+            decommitCount++;
         }
     }
+    while (++index < this->GetAvailablePageCount() && decommitCount < pageToDecommit);
+
+    if (index > startIndex)
```

Comment:
>index > startIndex [](start = 8, length = 18)

nit pick: continue to use "startIndex < index" the same as above. (Or move this logic and "pageCount = ..." into DecommitFreePagesInternal to avoid duplicates here.)